models/classes/runner/OfflineQtiRunnerService.php

```diff
@@ -0,0 +1,131 @@
+<?php
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA ;
+ */
+/**
+ * @author Péter Halász <peter@taotesting.com>
+ */
+
+namespace oat\taoQtiTest\models\runner;
+
+use oat\oatbox\service\ConfigurableService;
+use oat\taoQtiTest\models\runner\map\RunnerMap;
+
+/**
+ * Service class for the offline version of Qti Test Runner
+ */
+class OfflineQtiRunnerService extends ConfigurableService
+{
+    const SERVICE_ID = 'taoQtiTest/OfflineQtiRunnerService';
+
+    /**
+     * Returns an array of items, containing also confident data, like branching and response processing rules
+     *
+     * @param RunnerServiceContext $serviceContext
+     * @return array
+     * @throws \common_Exception
+     * @throws \common_exception_Error
+     * @throws \common_exception_InconsistentData
+     * @throws \common_exception_InvalidArgumentType
+     * @throws \tao_models_classes_FileNotFoundException
+     */
+    public function getItems($serviceContext)
+    {
+        $this->getRunnerService()->assertQtiRunnerServiceContext($serviceContext);
+
+        $runnerService = $this->getRunnerService();
+        $testMap = $runnerService->getTestMap($serviceContext);
+        $items = [];
+
+        foreach ($this->getItemIdentifiersFromTestMap($testMap) as $itemIdentifier) {
+            $itemRef = $runnerService->getItemHref($serviceContext, $itemIdentifier);
+
+            $items[$itemIdentifier] = [
+                'baseUrl' => $runnerService->getItemPublicUrl($serviceContext, $itemRef),
+                'itemData' => $runnerService->getItemData($serviceContext, $itemRef),
+                'itemState' => $runnerService->getItemState($serviceContext, $itemIdentifier),
+                'itemMetaData' => $runnerService->getFeedbacks($serviceContext, $itemRef),
```

Comment:
Is this key accurate? should not it be `feedbacks` instead?

Reply:
The `getFeedbacks` method returns the whole content of the `variableElements.json` file and as I saw there are other data than just the feedbacks. You're right, the naming is not correct, but not sure that I need to rename the method or the object key. What do you think?

Reply:
Both `feedbacks` and `itemMetaData` names aren't accurate and refer to something else. If this method returns the "variable elements" of an item (feedbacks, response delcaration and response processing, and conceptually anything that can help a test taker during the test), then why not naming it `getItemVariableElements` ? Or `getItemPrivateContent` , etc.

Regarding the response format, I don't think the `variableElements` should be sent separately from the `itemData`. Shouldn't they be merged merged upfront ? Remember those variable elements are only parts moved away from the original item definition. 
Then we can check if the response declaration is set or not (or we can have a flag).
What do you think ?

Reply:
I created a new function called `getItemVariableElementsData()` and marked the `getFeedbacks()` as deprecated. Also I mapped the variable elements into the `itemData` as you suggested. Could you please take a look?